pyro/contrib/forecast/evaluate.py

```diff
@@ -0,0 +1,163 @@
+    :param torch.Tensor truth: Ground truth.
+    :rtype: float
+    """
+    return crps_empirical(pred, truth).mean().cpu().item()
+
+
+DEFAULT_METRICS = {
+    "mae": eval_mae,
+    "rmse": eval_rmse,
+    "crps": eval_crps,
+}
+
+
+def backtest(data, covariates, model, *,
+             metrics=None,
+             transform=None,
+             train_window=None,
+             min_train_window=1,
+             test_window=None,
+             min_test_window=1,
+             stride=1,
+             seed=1234567890,
+             num_samples=100,
+             forecaster_options={}):
+    """
+    Backtest a forecasting model on a moving window of (train,test) data.
+
+    :param data: A tensor dataset with time dimension -2.
+    :type data: ~torch.Tensor
+    :param covariates: A tensor of covariates with time dimension -2.
+        For models not using covariates, pass a shaped empty tensor
+        ``torch.empty(duration, 0)``.
+    :type covariates: ~torch.Tensor
+    :param model:
+    :type model: ~pyro.contrib.forecast.forecaster.ForecastingModel
+
+    :param dict metrics: A dictionary mapping metric name to metric function.
+        The metric function should input a forecast ``pred`` and ground
+        ``truth`` and can output anything, often a number. Example metrics
+        include: :func:`eval_mae`, :func:`eval_rmse`, and :func:`eval_crps`.
+    :param callable transform: An optional transform to apply before computing
+        metrics. If provided this will be applied as
+        ``pred, truth = transform(pred, truth)``.
+    :param int train_window: Size of the training window. Be default trains
+        from beginning of data.
+    :param int min_train_window: If ``train_window`` is None, this specifies
+        the min training window size. Defaults to 1.
+    :param int test_window: Size of the test window. By default forecasts to
+        end of data.
+    :param int min_test_window: If ``test_window`` is None, this specifies
+        the min test window size. Defaults to 1.
+    :param int stride: Optional stride for test/train split. Defaults to 1.
+    :param int seed: Random number seed.
+    :param int num_samples: Number of samples for forecast.
+    :param dict forecaster_options: Options to pass to forecaster. See
+        :class:`~pyro.contrib.forecaster.Forecaster` for details.
+
+    :returns: A dict mapping (train_begin, split, test_end) triples to
+        dictionaries of evaluation data. Caller is responsible for aggregating
+        the per-window metrics.
+    :rtype: dict
+    """
+    assert data.size(-2) == covariates.size(-2)
+    assert isinstance(min_train_window, int) and min_train_window >= 1
+    assert isinstance(min_test_window, int) and min_test_window >= 1
+    if metrics is None:
+        metrics = DEFAULT_METRICS
+    assert metrics, "no metrics specified"
+
+    duration = data.size(-2)
+    if test_window is None:
+        stop = duration - min_test_window
```

Comment:
Should stop be `duration - min_test_window + 1` given that upper limit is exclusive?

Reply:
Good catch, fixed.